server/src-lib/Hasura/Server/Auth/WebHook.hs

```diff
@@ -0,0 +1,150 @@
+  res  <- liftIO $ try $ withAuthHook hook withGET withPOST
+  resp <- either logAndThrow return res
+  let status   = resp ^. Wreq.responseStatus
+      respBody = resp ^. Wreq.responseBody
+  mkUserInfoFromResp logger (ahUrl hook) (hookMethod hook) status respBody
+
+  where
+    withGET  url = Wreq.getWith  (mkOptions filteredHeaders) $ T.unpack url
+    withPOST url = Wreq.postWith (mkOptions [contentType]) (T.unpack url) $
+      object ["headers" J..= postHdrsPayload]
+    mkOptions = wreqOptions manager
+    contentType = ("Content-Type", "application/json")
+    postHdrsPayload = J.toJSON $ Map.fromList $ hdrsToText reqHeaders
+
+    logAndThrow err = do
+      unLogger logger $
+        WebHookLog LevelError Nothing (ahUrl hook) (hookMethod hook)
+        (Just $ HttpException err) Nothing Nothing
+      throw500 "Internal Server Error"
+
+    filteredHeaders = flip filter reqHeaders $ \(n, _) ->
+      n `notElem` commonClientHeadersIgnored
+
+
+-- | Processes the result of the wehook call, and tries to construct a
+--   'UserInfo'. The expiration time is retrieved from either the
+--   CacheControl or Expires header.
+mkUserInfoFromResp
+  :: (MonadIO m, MonadError QErr m)
+  => Logger Hasura
+  -> T.Text
+  -> N.StdMethod
+  -> N.Status
+  -> BL.ByteString
+  -> m (UserInfo, Maybe UTCTime)
+mkUserInfoFromResp (Logger logger) url method statusCode respBody
+  | statusCode == N.status200 =
+    case eitherDecode respBody of
+      Left e -> do
+        logError
+        throw500 $ "Invalid response from authorization hook: " <> T.pack e
+      Right rawHeaders -> getUserInfoFromHdrs rawHeaders
+
+  | statusCode == N.status401 = do
+    logError
+    throw401 "Authentication hook unauthorized this request"
+
+  | otherwise = do
+    logError
+    throw500 "Invalid response from authorization hook"
+  where
+    getUserInfoFromHdrs rawHeaders = do
+      let usrVars = mkUserVars $ Map.toList rawHeaders
+      case roleFromVars usrVars of
+        Nothing -> do
+          logError
+          throw500 "missing x-hasura-role key in webhook response"
+        Just rn -> do
+          logWebHookResp LevelInfo Nothing Nothing
+          let ccHeader = Map.lookup "Cache-Control" rawHeaders
+              eHeader  = Map.lookup "Expires"       rawHeaders
+          expTime <- getExpiryTime ccHeader eHeader
+          return (mkUserInfo rn usrVars, expTime)
+
+    logWarn =
+      logWebHookResp LevelWarn $ Just respBody
+    logError =
+      logWebHookResp LevelError (Just respBody) Nothing
+    logWebHookResp logLevel mResp message =
+      logger $ WebHookLog logLevel (Just statusCode)
+        url method Nothing (bsToTxt . BL.toStrict <$> mResp) message
+
+    getExpiryTime ccHeader eHeader = do
+      cct <- timeFromCacheControlHeader ccHeader $ logWarn . Just
+      expTime <- case cct of
+        Just t  -> return $ Just t
+        Nothing -> timeFromExpiresHeader eHeader $ logWarn . Just
+      onMaybe expTime $ fmap Just . toUTCTime
```

Comment:
You might be able to get more bang for your buck by using `MaybeT`’s `Alternative` instance here. You could do something like this:

```haskell
mkUserInfoFromResp =
    ...
    expiration <- runMaybeT (timeFromCacheControl headers <|> timeFromExpires headers)
    ...
  where
    timeFromCacheControl headers = do
      header <- afold $ Map.lookup "Cache-Control" headers
      duration <- parseMaxAge header `onLeft` \err -> logWarn err *> empty
      addUTCTime duration <$> liftIO getCurrentTime

    timeFromExpires headers = do
      header <- afold $ Map.lookup "Expires" headers
      timeFromExpiresHeader header `onLeft` \err -> logWarn err *> empty

    -- I think this function should go in Hasura.Prelude; I’ve wanted it on several occasions.
    afold :: (Foldable t, Alternative f) => t a -> f a
    afold = getAlt . foldMap pure
```

Reply:
Done!